### PR TITLE
MPD and DATA extension support

### PR DESCRIPTION
**Pull request by ScanMountGoat**
This PR adds support for documents with multiple files (MPD) as well as adding the parsing for embedded data files. This adds what I would consider the "bare minimum" functionality for working with LDraw files in an application like Blender. The main functionality missing is the BFC extension and color handling.

weldr lib
* Calling `parse` returns the main file for MPD files or the entire root file otherwise. This avoids loading some submodel geometry multiple times for MPD files.
* When loading a `SourceFile` in the `parse` function, the file is assumed to be an MPD file and split into multiple files. If the split file list is empty, the entire file is used to still handle dat and ldr files normally.
* The `parse` function returns a `SourceFile` instead of a reference. This simplifies the implementation but uses an additional clone. The old implementation avoided lifetime issues but didn't guarantee that the reference was valid. The return type can be changed later as long as it indicates the "main model" of the file.

weldr bin
* The hierarchy of submodels is preserved when converting to gltf. This matches the behavior of other import/export tools when working with MPD files.
* Meshes for parts are cached and instanced, resulting in lower memory usage and faster export times. This assumes that all .dat files are parts for now. This can be replaced by checking the type like "part" or "subpart" in the future.
* Forward and backward slashes are tried using `path-slash` to fix some resolve errors in Linux and MacOS.
* Prevents some validation warnings from the gltf Validator and VSCode extension. These errors mostly impact MPD files. I've noticed that some implementations of gltf like Blender will fail with strange errors if the file has validation warnings.

### CODE REVIEW

**Comment by ScanMountGoat** [comment]
The UCS Falcon loaded into Blender from gltf after adding some unofficial part paths to the resolver.
image

File: lib/Cargo.toml [hunk 1]
```diff
@@ -29,3 +29,4 @@ path = "src/lib.rs"
 nom = "7.1.3"
 cgmath = "0.17.0"
 log = "0.4"
+base64 = "0.21.0"
```
Comment (djeedai) [inline lib/Cargo.toml:32]: This needs to bump min Rust supported version to 1.57, which is fine I think but needs to be done as part of that PR otherwise CI fails. See 8c7c6762f4c87ee2772c878bd9275cf930bbfbb0 for a model of where things need to be changed. Thanks!

File: lib/src/parse.rs [hunk 2]
```diff
@@ -0,0 +1,1299 @@
+use base64::Engine;
```
Comment (djeedai) [inline lib/src/parse.rs:1]: This kind of moving things around makes the PR quite hard to compare, as diffs won't work. Could you please revert? We can move things around with a later clean-up, but now I can't tell what changed. Thanks!

File: bin/weldr/src/weldr.rs [hunk 3]
```diff
@@ -225,24 +223,22 @@ impl DiskResolver {
...
+        self.base_paths
+            .iter()
+            .find_map(|prefix| {
+                // The file's path separator may not match the current OS.
+                // Don't assume the current file system and try both separators.
+                let forward_path = prefix.join(PathBuf::from_slash(filename));
+                let backward_path = prefix.join(PathBuf::from_backslash(filename));
+                std::fs::read(prefix.join(forward_path))
+                    .or_else(|_| std::fs::read(prefix.join(backward_path)))
+                    .ok()
```
Comment (djeedai) [inline bin/weldr/src/weldr.rs]: Not a big fan of that. The separator for a given OS is well-known, so trying the other character which is not a separator has a potential to craft either invalid paths or paths to other unintended locations, which has implications for integrity/security. I'd rather test for the particular OS path, since the purpose of the `join()` is for use with an OS call below. And if the issue here is that `prefix` itself may contain the wrong kind of separator, then I'm in favor of always using `/` only internally, and converting to OS-specific separator at the last minute when doing an OS call. And of course fail when assigning `prefix` from a user value that contains `\`.

File: bin/weldr/src/convert.rs [hunk 4]
```diff
@@ -107,154 +103,261 @@ impl ConvertCommand {
...
+    fn add_nodes(
+        &self,
+        source_file: &weldr::SourceFile,
+        transform: weldr::Mat4,
+        source_map: &weldr::SourceMap,
+        gltf: &mut gltf::Gltf,
+        buffer: &mut Vec<u8>,
+        mesh_cache: &mut HashMap<String, Option<usize>>,
+    ) -> u32 {
+        let matrix = if transform == weldr::Mat4::from_scale(1.0) {
```
Comment (djeedai) [inline bin/weldr/src/convert.rs]: I'd rather take `transform: Option<weldr::Mat4>` directly instead of that floating point comparison. Especially as we're calling below `add_nodes()` with `Mat4::from_scale(1.0)` so building a new identity matrix just for the sake of comparing it against another identity matrix we build here and discard it. We could just pass `None` and use as-is inside this method.

File: bin/weldr/src/convert.rs [hunk 5]
```diff
@@ -107,154 +103,261 @@ impl ConvertCommand {
...
+        let node_index = gltf.nodes.len();
+        let node = gltf::Node {
+            name: Some(source_file.filename.clone()),
+            children: Vec::new(),
+            mesh_index: None,
+            matrix,
+        };
+        gltf.nodes.push(node);
+
+        // TODO: Check the part type rather than the extension.
```
Comment (djeedai) [inline bin/weldr/src/convert.rs:127]: What amount of work is it to fix that now? Extension-based guesses is a bit ugly.

File: bin/weldr/src/convert.rs [hunk 6]
```diff
@@ -107,154 +103,261 @@ impl ConvertCommand {
...
+        if source_file.filename.ends_with(".dat") {
+            // Create geometry if the node is a part.
+            let mesh_index = mesh_cache
+                .entry(source_file.filename.clone())
+                .or_insert_with(|| {
+                    let mesh_index = gltf.meshes.len();
+                    let geometry = self.create_geometry(source_file, source_map);
+                    // Don't set empty meshes to avoid import errors.
+                    if !geometry.vertices.is_empty() && !geometry.triangle_indices.is_empty() {
+                        self.add_mesh(&geometry, gltf, buffer);
```
Comment (djeedai) [inline bin/weldr/src/convert.rs]: It's slightly weird to allocate a separate `GeometryCache` object, then check if empty, then copy the data again into a separate buffer and discard that object. You should be able to directly insert into the final buffer, and early-out if no insertion occured, no?

File: bin/weldr/src/convert.rs [hunk 7]
```diff
@@ -107,154 +103,261 @@ impl ConvertCommand {
...
+                        Some(mesh_index)
+                    } else {
+                        None
+                    }
+                });
+
+            gltf.nodes[node_index].mesh_index = mesh_index.map(|i| i as u32);
```
Comment (djeedai) [inline bin/weldr/src/convert.rs]: You can directly return `Some(mesh_index as u32)` above, that would avoid that `map()` no?

File: bin/weldr/src/convert.rs [hunk 8]
```diff
@@ -107,154 +103,261 @@ impl ConvertCommand {
...
+                if let Command::SubFileRef(sfr_cmd) = cmd {
+                    if let Some(subfile) = source_map.get(&sfr_cmd.file) {
+                        // Don't apply node transforms to preserve the scene hierarchy.
+                        // Applications should handle combining the transforms.
+                        let transform = weldr::Mat4::from_cols(
+                            weldr::Vec4::new(sfr_cmd.row0.x, sfr_cmd.row1.x, sfr_cmd.row2.x, 0.0),
+                            weldr::Vec4::new(sfr_cmd.row0.y, sfr_cmd.row1.y, sfr_cmd.row2.y, 0.0),
+                            weldr::Vec4::new(sfr_cmd.row0.z, sfr_cmd.row1.z, sfr_cmd.row2.z, 0.0),
+                            weldr::Vec4::new(sfr_cmd.pos.x, sfr_cmd.pos.y, sfr_cmd.pos.z, 1.0),
+                        );
```
Comment (djeedai) [inline bin/weldr/src/convert.rs]: FYI logged #28 because this is quite ugly. Ignoring for now though.

File: bin/weldr/src/convert.rs [hunk 9]
```diff
@@ -107,154 +103,261 @@ impl ConvertCommand {
...
+        let mut gltf = gltf::Gltf {
+            asset,
+            nodes: Vec::new(),
+            scenes: vec![scene],
+            buffers: Vec::new(),
+            buffer_views: Vec::new(),
+            accessors: Vec::new(),
+            meshes: Vec::new(),
+            scene: Some(0),
         };
```
Comment (djeedai) [inline bin/weldr/src/convert.rs:194]: Out of scope but this type should probably derive `Default` and we'd use `..Default::default()` here.

File: bin/weldr/src/convert.rs [hunk 10]
```diff
@@ -107,154 +103,261 @@ impl ConvertCommand {
...
+                .map(|v| [v.x, v.y, v.z]),
+        };
+
+        let mut primitives = Vec::new();
+        // TODO: Line indices.
+        if !geometry_cache.triangle_indices.is_empty() {
+            let attributes = HashMap::from([("POSITION".to_string(), gltf.accessors.len() as u32)]);
+            gltf.accessors.push(vertex_accessor);
+
+            // TODO: Use bytemuck instead.
```
Comment (djeedai) [inline bin/weldr/src/convert.rs:299]: Agreed. Logged as #29.

File: lib/src/lib.rs [hunk 11]
```diff
@@ -1280,738 +737,135 @@ pub trait FileRefResolver {
...
-        assert_eq!(
-            keywords_list(b""),
-            Err(nom_error(&b""[..], ErrorKind::TakeWhile1))
-        );
-        assert_eq!(keywords_list(b"a"), Ok((&b""[..], vec!["a"])));
-        assert_eq!(keywords_list(b"a,b,c"), Ok((&b""[..], vec!["a", "b", "c"])));
-    }
-
-    #[test]
-    fn test_filename_char() {
```
Comment (djeedai) [inline lib/src/lib.rs:1783]: Why was this test deleted?

File: lib/src/lib.rs [hunk 12]
```diff
@@ -98,109 +81,12 @@ pub type Mat4 = cgmath::Matrix4<f32>;
...
-        Command::Keywords(KeywordsCmd {
-            keywords: keywords.iter().map(|kw| kw.trim().to_string()).collect(),
-        }),
-    ))
-}
+pub use error::{Error, ParseError, ResolveError};
 
 /// RGB color in sRGB color space.
-#[derive(Debug, PartialEq)]
+#[derive(Debug, PartialEq, Clone)]
```
Comment (djeedai) [inline lib/src/lib.rs]:
```suggestion
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
```

File: lib/src/lib.rs [hunk 13]
```diff
@@ -855,77 +305,79 @@ fn load_and_parse_single_file(
...
-    resolver: &dyn FileRefResolver,
+    resolver: &R,
     source_map: &mut SourceMap,
-) -> Result<SourceFileRef, Error> {
-    if let Some(existing_file) = source_map.find_filename(filename) {
-        return Ok(existing_file);
+) -> Result<SourceFile, Error> {
+    // TODO: Avoid clone.
+    if let Some(existing_file) = source_map.get(filename) {
+        // TODO: Caching is handled here, so the disk resolver doesn't need to cache?
```
Comment (djeedai) [inline lib/src/lib.rs]: There's no disk resolver in the library. There shouldn't be any assumption about the resolver being disk-based or otherwise, nor doing or not any caching.

File: lib/src/lib.rs [hunk 14]
```diff
@@ -855,77 +305,79 @@ fn load_and_parse_single_file(
...
+        // TODO: Is this fact documented?
+        // TODO: If we have the SourceMap, can't we just return an actual reference?
```
Comment (djeedai) [inline lib/src/lib.rs]: Isn't that exactly what the original code was doing? The point of returning a `SourceFileRef` was that if you try to load a second time you get the same reference as previously, instead of getting a cloned `SourceFile`. I'm not sure I like that new approach.

File: lib/src/lib.rs [hunk 15]
```diff
@@ -855,77 +305,79 @@ fn load_and_parse_single_file(
...
 ///   let mut source_map = SourceMap::new();
-///   let root_file_ref = parse("root.ldr", &resolver, &mut source_map)?;
-///   let root_file = root_file_ref.get(&source_map);
+///   let root_file = parse("root.ldr", &resolver, &mut source_map)?;
 ///   assert_eq!(root_file.filename, "root.ldr");
 ///   Ok(())
 /// }
 /// ```
-pub fn parse(
+pub fn parse<'a, R: FileRefResolver>(
```
Comment (djeedai) [inline lib/src/lib.rs]: What's this `'a` lifetime?

File: lib/src/lib.rs [hunk 16]
```diff
@@ -855,77 +305,79 @@ fn load_and_parse_single_file(
...
             Some(_) => trace!("Already parsed; reusing sub-file: {}", filename),
             None => {
                 trace!("Not yet parsed; parsing sub-file: {}", filename);
-                let source_file = load_and_parse_single_file(&filename[..], resolver)?;
-                let source_file_ref = source_map.insert(source_file);
+                load_file(filename, resolver, source_map, &mut stack)?;
                 trace!(
                     "Post-loading resolving subfile refs of sub-file: {}",
                     filename
                 );
```
Comment (djeedai) [inline lib/src/lib.rs]: This is a trace for the line below that was deleted. So it makes few sense, should probably also be deleted.

File: lib/src/lib.rs [hunk 17]
```diff
@@ -1007,171 +459,164 @@ pub struct ColourCmd {
...
                 }
             }
         }
-
-        // Restore the filename map
-        std::mem::swap(&mut filename_map, &mut self.filename_map);
     }
 }
 
+fn split_files(commands: &[Command]) -> Vec<SourceFile> {
```
Comment (djeedai) [inline lib/src/lib.rs]: This is called "split" but will return an empty `Vec` if this is not an MPD file, right? So it's only splitting MPD files, not regular ones. Can we rename to clarify, _e.g._ `split_mpd_file`?

File: lib/src/lib.rs [hunk 18]
```diff
@@ -1007,171 +459,164 @@ pub struct ColourCmd {
...
-        }
-    }
-}
-
-impl SourceFileRef {
-    pub fn get<'a>(&'a self, source_map: &'a SourceMap) -> &'a SourceFile {
-        source_map.get(*self)
-    }
+    /// Map of filenames to source files.
+    // TODO: How to handle case sensitivity like STUD.DAT vs stud.dat?
```
Comment (djeedai) [inline lib/src/lib.rs]: With a newtype `SourceFileRef(String)` and a custom implementation of `Hash` I guess? But that's probably more complicated than just always converting any entry to lowercase for example and compare like that.

File: lib/src/lib.rs [hunk 19]
```diff
@@ -1007,171 +459,164 @@ pub struct ColourCmd {
...
     /// Comment content, excluding the command identififer `0` and the optional comment marker `//`.
     pub text: String,
 }
 
+impl CommentCmd {
+    pub fn new(text: &str) -> CommentCmd {
+        CommentCmd {
+            text: text.to_string(),
+        }
+    }
```
Comment (djeedai) [inline lib/src/lib.rs:400]: nit:

```suggestion
    pub fn new(text: impl Into<String>) -> CommentCmd {
        CommentCmd {
            text: text.into(),
        }
    }
```

File: lib/src/lib.rs [hunk 20]
```diff
@@ -1007,171 +459,164 @@ pub struct ColourCmd {
...
-            .map(|&index| SourceFileRef { index })
-    }
-
-    /// Insert a new source file into the collection.
-    fn insert(&mut self, source_file: SourceFile) -> SourceFileRef {
-        if let Some(&index) = self.filename_map.get(&source_file.filename) {
-            SourceFileRef { index }
+    /// Inserts a new source file into the collection.
+    /// Returns a copy of `source_file` or the main file for multi-part documents (MPD).
+    pub fn insert(&mut self, source_file: SourceFile) -> SourceFile {
```
Comment (djeedai) [inline lib/src/lib.rs]: That cloning thing again is quite ugly and inefficient. I'd rather return a reference to the file into the map, which can be either a simple `String` for now, or a newtype which normalized the key of the `HashMap` as discussed in another comment on casing. Then if the user is really interested in retrieving the source file (that they just inserted!) they can use `get()` or `get_mut()`, but there's a good chance the code inserting doesn't need to query immediately.

File: lib/tests/parse.rs [hunk 21]
```diff
@@ -1,26 +1,9 @@
 extern crate weldr;
 
-use weldr::{error::ResolveError, Command, FileRefResolver, SourceFileRef, SourceMap, SubFileRef};
+use weldr::{error::ResolveError, Command, FileRefResolver, SourceMap, SubFileRefCmd};
 
 use std::collections::HashMap;
 
-fn print_rec(source_map: &SourceMap, source_file_ref: SourceFileRef, indent: usize) {
```
Comment (djeedai) [inline lib/tests/parse.rs:7]: Can we leave this, if only for debugging? This is quite useful if I remember correctly.

**Review by djeedai — CHANGES_REQUESTED** [review]
I quite like the direction this is taking. Things like the removal of `SubFileRef` and using direct `String` references looks much simpler. There's a few things however I'm not convinced about, see individual comments for details. The blockers are probably few though, like the minimum Rust version bump. Thanks!

File: lib/src/parse.rs (hunk 2 above)
Comment (djeedai) [reply lib/src/parse.rs:1]: Actually nevermind I reviewed all already.

File: lib/src/lib.rs (hunk 18 above)
Comment (ScanMountGoat) [reply lib/src/lib.rs]: That assumes the LDraw parts library uses a consistent casing for filenames. It's worth making a separate issue for this, since I'm not sure if there's a "correct" way of handling it. If the command references "STUD.DAT" but the file system has both "stud.dat" and "Stud.DAT", which file do we pick?

File: lib/src/lib.rs (hunk 20 above)
Comment (ScanMountGoat) [reply lib/src/lib.rs]: Changed to just return `String`

File: lib/src/lib.rs (hunk 11 above)
Comment (ScanMountGoat) [reply lib/src/lib.rs:1783]: The "filenames" can also be string identifiers for submodels in MPD files.

File: bin/weldr/src/convert.rs (hunk 9 above)
Comment (ScanMountGoat) [reply bin/weldr/src/convert.rs:194]: It may also be better in the future to just use https://crates.io/crates/gltf

File: bin/weldr/src/convert.rs (hunk 5 above)
Comment (ScanMountGoat) [reply bin/weldr/src/convert.rs:127]: There's a special comment command for determining if the file is an official or unofficial "part". This should also be updated to check if each file contains geometry commands since apparently some files embed geometry inline for things like hoses and tubes. I'm not sure how trivial this would be to implement. Calling `root_file.iter(...)` can create enough geometry to crash many applications, so at least some level of instancing is necessary for medium to large models.

File: lib/src/lib.rs (hunk 18 above)
Comment (ScanMountGoat) [reply lib/src/lib.rs]: Apparently "STUD.DAT" is a weird special case. I've created https://github.com/djeedai/weldr/issues/30 to track this.

File: bin/weldr/src/convert.rs (hunk 5 above)
Comment (djeedai) [reply bin/weldr/src/convert.rs:127]: Ok I think I didn't understand then what is happening here. Why do we need to check for the official-ness of the part, and how is that related with the file extension?

File: bin/weldr/src/convert.rs (hunk 9 above)
Comment (djeedai) [reply bin/weldr/src/convert.rs:194]: Logged as #31.

File: lib/src/lib.rs (hunk 19 above)
Comment (djeedai) [reply lib/src/lib.rs:400]: I still see the old code, is that resolved? Or did I get confused by the GitHub once again? :)

File: lib/src/lib.rs (hunk 18 above)
Comment (djeedai) [reply lib/src/lib.rs]: > That assumes the LDraw parts library uses a consistent casing for filenames.

You mean my suggestion? No it's not, it's working around that:

```rust
struct SourceFileRef(String);

impl SourceFileRef {
  fn new(str: impl Into<String>) -> Self {
    let str = str.into();
    Self(str.to_lowercase())
  }
}

impl Hash for SourceFileRef {
  fn hash<H: Hasher>(&self, state: &mut H) {
    self.0.hash(state);
  }
}

impl PartialEq for SourceFileRef {
  fn eq(&self, other: &Self) -> bool {
    self.0 == other.0
  }
}

impl Eq for SourceFileRef {}
```

> If the command references "STUD.DAT" but the file system has both "stud.dat" and "Stud.DAT", which file do we pick?

There's no solution. The specification says it should be case-insensitive, so if that situation happens the scenario itself is ill-defined and there is no unique solution:
- you could decide to pick one randomly, assuming they're exact duplicates, but that may produce inconsistent results if they're not.
- you could decide to try and find an exact case-sensitive match, assuming the 2 files are different, but that goes against the specification.

I think we should just ignore that case. If the user or calling code put itself in that situation they shouldn't expect us to solve it without their help. So just doing all comparisons case-insensitive should be enough to handle all well-formed usages.

> Apparently "STUD.DAT" is a weird special case. I've created #30 to track this.

The forum you linked has a few posts explicitly saying "please do ignore this, it's undocumented and nobody knows about it except `ldraw.exe`, you should absolutely ignore that rule about STUD.DAT". I agree with their analysis.

File: lib/src/lib.rs (hunk 11 above)
Comment (djeedai) [reply lib/src/lib.rs:1783]: I don't understand. This test is aimed at validating `filename_char()`. Since that parser is still in use, why delete its test?

**Review by djeedai — COMMENTED** [review]
Thanks for the changes. There's a couple more things I'm not sure I understand (see comments) but that's otherwise roughly in a mergeable state. Thanks!

File: bin/weldr/src/convert.rs (hunk 5 above)
Comment (ScanMountGoat) [reply bin/weldr/src/convert.rs:127]: There's more than one type that determines if it is a part. Ideally, we could just create a mesh for each ldraw file. In practice, this creates an unmanageable number of objects in the scene when importing. Stopping the recursion at the "part" level uses more memory but makes the file easier to work with. Whether the overhead matters depends on the application.

**Comment by ScanMountGoat** [comment]
The filename_char code has been completely removed since it's not actually used anywhere. The spec encourages not using special characters in filenames but doesn't require it. Applying validation beyond just checking for utf8 won't work with MPD files since the files don't need to be files on disk. These will be strings entered by the user in applications like Studio, LeoCAD, etc.

**Comment by ScanMountGoat** [comment]
I've adjusted the path normalization to only use the OS specific separator. This also seems to be the approach used by LDView. This also uses glam to simplify some of the matrix math. Let me know if there are any issues blocking this PR.

File: bin/weldr/src/convert.rs (hunk 5 above)
Comment (djeedai) [reply bin/weldr/src/convert.rs:127]: If I understand correctly you're saying that:
- Parsing and building a mesh for each single file in a document produces too many objects for an application to handle.
- Building a single "merged" mesh for the entire document conversely creates a mesh too big to handle.
- The middle ground is to create a smaller quantity of large-ish meshes, and for that we use a heuristic here by stopping the recursion at the part level, by guessing what is a reusable part (instancing).

Did I get the reasoning correctly?

File: bin/weldr/src/weldr.rs [hunk 22]
```diff
@@ -195,7 +192,6 @@ impl DiskResolver {
             .map_err(|e| Error::NotFound(format!("catalog path not found ({})", e)))?;
         let base_paths = vec![
             catalog_path.join("p"),
-            catalog_path.join("p").join("48"),
```
Comment (djeedai) [inline bin/weldr/src/weldr.rs:198]: Why was this removed? Is the latest catalog not using a "48" folder anymore?

**Review by djeedai — COMMENTED** [review]
I'd just like to understand what looks like a heuristic to balance mesh size vs. object count (?), and after that it's good to go!

File: bin/weldr/src/weldr.rs (hunk 22 above)
Comment (ScanMountGoat) [reply bin/weldr/src/weldr.rs:198]: > Why was this removed? Is the latest catalog not using a "48" folder anymore?

The 48 folder is the high resolution primitives folder. There are also low res primitives in the 8 folder. Normal resolution primitives are just in the p folder. Some applications support switching primitives. Just sticking with standard resolution primitives seems like a reasonable default for gltf export for now. Application code can always write their own resolver.

> Did I get the reasoning correctly?

Yeah. It's a tradeoff between memory usage and per object overhead. The extension check is a bit of a hack, but I haven't run into any issues doing it that way while developing https://github.com/ScanMountGoat/ldr_tools_blender. It's probably worth revisiting at some point.

Older LDraw applications use immediate mode rendering on each triangle or quad line in each file with push/pop matrix commands. This made sense on older hardware with limited memory but performs poorly even on newer GPUs. All of the modern applications and importers I've looked at merge all the geometry for each "part". 

A more important reason to merge like this is that apparently subparts sometimes use shearing matrices. This isn't a problem for gltf but won't work in applications that decompose transforms into translation, rotation, and scale. This can be resolved by merging all the geometry at the part level. You can see an example if you search for "shear" on the readme of https://github.com/cuddlyogre/ExportLDraw.

**Review by djeedai — APPROVED** [review]

**Comment by djeedai** [comment]
Thanks a lot for that fantastic contribution and pushing through all the comments @ScanMountGoat, much appreciated!
